### ISSUE TITLE
Incorrect halo update scheduling

### ISSUE THREAD

**Issue by mloubout** [devitocodes/devito#1459]
The halo update is incorrectly placed when using a mix of time and not time varying object

MFE:

```
from devito import Eq, Operator, VectorFunction, div, grad, Grid, TimeFunction

grid = Grid((11, 11))

s = grid.stepping_dim.spacing
st = tuple([None]*grid.dim)

p = TimeFunction(name="p", grid=grid, time_order=2, space_order=4)
tmp = VectorFunction(name="tmp", grid=grid, space_order=4, staggered=st)


op = Operator([Eq(tmp, grad(p)), Eq(p.forward, div(tmp))])

```

This produces the following code:

```
    haloupdate_0(p_vec,comm,nb,t0,nthreads);
    haloupdate_1(tmp_x_vec,comm,nb,nthreads);
    haloupdate_2(tmp_y_vec,comm,nb,nthreads);
    #pragma omp parallel num_threads(nthreads)
    {
      #pragma omp for collapse(1) schedule(dynamic,1)
      for (int x = x_m; x <= x_M; x += 1)
      {
        #pragma omp simd aligned(p,tmp_x,tmp_y:16)
        for (int y = y_m; y <= y_M; y += 1)
        {
          tmp_x[x + 4][y + 4] = (8.33333333e-2F*(p[t0][x + 2][y + 4] - p[t0][x + 6][y + 4]) + 6.66666667e-1F*(-p[t0][x + 3][y + 4] + p[t0][x + 5][y + 4]))/h_x;
          tmp_y[x + 4][y + 4] = (8.33333333e-2F*(p[t0][x + 4][y + 2] - p[t0][x + 4][y + 6]) + 6.66666667e-1F*(-p[t0][x + 4][y + 3] + p[t0][x + 4][y + 5]))/h_y;
        }
      }
    }
    #pragma omp parallel num_threads(nthreads)
    {
      #pragma omp for collapse(1) schedule(dynamic,1)
      for (int x = x_m; x <= x_M; x += 1)
      {
        #pragma omp simd aligned(p,tmp_x,tmp_y:16)
        for (int y = y_m; y <= y_M; y += 1)
        {
          p[t1][x + 4][y + 4] = (8.33333333e-2F*(tmp_y[x + 4][y + 2] - tmp_y[x + 4][y + 6]) + 6.66666667e-1F*(-tmp_y[x + 4][y + 3] + tmp_y[x + 4][y + 5]))/h_y + (8.33333333e-2F*(tmp_x[x + 2][y + 4] - tmp_x[x + 6][y + 4]) + 6.66666667e-1F*(-tmp_x[x + 3][y + 4] + tmp_x[x + 5][y + 4]))/h_x;
        }
      }
    }

```

Which is incorrect,     `haloupdate_1(tmp_x_vec,comm,nb,nthreads)` and `haloupdate_2(tmp_y_vec,comm,nb,nthreads)` should be called after the `tmp` loop because `p[t1]` needs to compute FD of it.

**Comment by Leitevmd** [comment]
Hey guys, 

My name is Victor, I'm working with @nogueirapeterson on the SENAI-CIMATEC team. 

I was looking into that question and found this behavior is introduced by rule2 within [_merge_halospots(iet)](https://github.com/devitocodes/devito/blob/master/devito/passes/iet/mpi.py#L146) which is called in [optimize_halospots(iet)](https://github.com/devitocodes/devito/blob/master/devito/passes/iet/mpi.py#L16). I was able to overcome this by omitting rule2 in line [151](https://github.com/devitocodes/devito/blob/master/devito/passes/iet/mpi.py#L151) of mpi.py:
```
# merge_rules = [rule0, rule1, rule2]
merge_rules = [rule0, rule1]
```

I'm not sure yet what this rule exactly means, so I'm not sure if this is a misbehavior due to other code or to a badly formulated rule. If you guys bring some light into that question I'm here to help. 

Hope this helps in some way.

Cheers

**Comment by FabioLuporini** [comment]
@Leitevmd congratz on getting deep down to that point -- yes, that's potentially the source of the problem.

As you may imagine, however, we cannot just comment out `rule2`; if you do so and try running all tests in `test_mpi.py`, you'll see that some of them will start failing. But you're definitely looking in the right direction. You need to either fix one of the three rules (maybe `rule2` itself?) or add another rule such that the halo spots don't get merged, and so eventually they appear at the right place in the generated code

**Comment by Leitevmd** [comment]
@FabioLuporini, `rule2` always return `True` if `loc_indices == {}` because of its return statement: ([mpi.py#L146](https://github.com/devitocodes/devito/blob/master/devito/passes/iet/mpi.py#L146))
```
    def rule2(dep, hs, loc_indices):
        # E.g., `dep=W<f,[t1, x+1]> -> R<f,[t1, xl+1]>` and `loc_indices={t: t0}` => True
        return all(dep.distance_mapper[d] == 0 and dep.source[d] is not v
                   for d, v in loc_indices.items())
```
I bet this is not what this rule wants. Adding `and loc_indices != {}` at the end of the return statement prevents the undesired merge of halo spots reported on this issue and I think it also maintains the proper purpose of `rule2`. I'm assuming this because the only test in `test_mpi.py` which fails when I simply omit `rule2` is `test_avoid_haloudate_if_flowdep_along_other_dim` ([test_mpi.py#L914](https://github.com/devitocodes/devito/blob/master/tests/test_mpi.py#L914)). With the change mentioned above all tests in `test_mpi.py` pass. What you think, could this be a reasonable solution?

**Comment by FabioLuporini** [comment]
I think you're looking in the right direction.

What if we instead turn the `all` into an `any` ?

I think it makes it a bit neater while achieving the same purpose. Also, `any` logically makes more sense to me as we only need one access index (eg `t1`) to differ from the loc-index (eg `t0`) such that this dependence doesn't represent a treat to the halo merge. 

Does it work? 

I suggest you open a PR and add the MFE that @mloubout attached to this issue. Please, place the MFE right below `test_avoid_haloudate_if_flowdep_along_other_dim ` and give it a suitable name.

**Comment by Leitevmd** [comment]
@FabioLuporini  I've opend a PR  (#1482)  with the solution you suggested but instead of using the MFE above I've wrote a simpler test (in accordance with other tests within test_mpi.py) with minimal requirements to reproduce this issue.

**Comment by FabioLuporini** [comment]
@Leitevmd nice job, and I really do like the new MFE !

I left a couple of minor comments in the PR. Happy to merge once they're addressed

**Comment by Leitevmd** [comment]
@FabioLuporini I think, they're all addressed. I have also included a counting check for halo update calls (like previous tests).

**Comment by FabioLuporini** [comment]
awesome, gonna merge soon, then will close this issue.

OT: BTW, are u on our slack channels @Leitevmd ? you should be :)

**Comment by FabioLuporini** [comment]
fixed via #1482 , hence closing
